// yamdbf/bin/lib/command/base/groupcontrol/DisableGroup.js
'use babel';
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _Command = require('../../Command');

var _Command2 = _interopRequireDefault(_Command);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

Function.prototype.$asyncbind = function anonymous(self, catcher) {
	var resolver = this;

	if (catcher === true) {
		if (!Function.prototype.$asyncbind.EagerThenable) Function.prototype.$asyncbind.EagerThenable = function factory(tick) {
			var _tasks = [];

			if (!tick) {
				try {
					tick = process.nextTick;
				} catch (ex) {
					tick = function tick(p) {
						setTimeout(p, 0);
					};
				}
			}

			function _untask() {
				for (var i = 0; i < _tasks.length; i += 2) {
					var t = _tasks[i + 1],
					    r = _tasks[i];

					for (var j = 0; j < t.length; j++) t[j].call(null, r);
				}

				_tasks = [];
			}

			function isThenable(obj) {
				return obj && obj instanceof Object && typeof obj.then === 'function';
			}

			function EagerThenable(resolver) {
				function done(inline) {
					var w;
					if (_sync || phase < 0 || (w = _thens[phase]).length === 0) return;

					_tasks.push(result, w);

					_thens = [[], []];
					if (_tasks.length === 2) inline ? _untask() : tick(_untask);
				}

				function resolveThen(x) {
					if (phase >= 0) return;
					if (isThenable(x)) return x.then(resolveThen, rejectThen);
					phase = 0;
					result = x;
					done(true);
				}

				function rejectThen(x) {
					if (phase >= 0) return;
					if (isThenable(x)) return x.then(resolveThen, rejectThen);
					phase = 1;
					result = x;
					done(true);
				}

				function settler(resolver, rejecter) {
					_thens[0].push(resolver);

					_thens[1].push(rejecter);

					done();
				}

				function toString() {
					return 'EagerThenable{' + {
						'-1': 'pending',
						0: 'resolved',
						1: 'rejected'
					}[phase] + '}=' + result.toString();
				}

				function guard() {
					try {
						resolver.call(null, resolveThen, rejectThen);
					} catch (ex) {
						rejectThen(ex);
					}
				}

				this.then = settler;
				this.toString = toString;
				var _thens = [[], []],
				    _sync = true,
				    phase = -1,
				    result;
				guard();
				_sync = false;
				done();
			}

			EagerThenable.resolve = function (v) {
				return isThenable(v) ? v : {
					then: function then(resolve, reject) {
						return resolve(v);
					}
				};
			};

			return EagerThenable;
		}();
		return new Function.prototype.$asyncbind.EagerThenable(boundThen);
	}

	if (catcher) {
		if (Function.prototype.$asyncbind.wrapAsyncStack) catcher = Function.prototype.$asyncbind.wrapAsyncStack(catcher);
		return then;
	}

	function then(result, error) {
		try {
			return result && result instanceof Object && typeof result.then === 'function' ? result.then(then, catcher) : resolver.call(self, result, error || catcher);
		} catch (ex) {
			return (error || catcher)(ex);
		}
	}

	function boundThen(result, error) {
		return resolver.call(self, result, error);
	}

	boundThen.then = boundThen;
	return boundThen;
};

class DisableGroup extends _Command2.default {
	constructor(bot) {
		super(bot, {
			name: 'disablegroup',
			description: 'Disable a command group',
			aliases: ['disable', 'dg'],
			usage: '<prefix>disablegroup <group>',
			extraHelp: 'Disables a command group so that all of the commands in the group cannot be used on this server.',
			group: 'base',
			permissions: ['ADMINISTRATOR']
		});
	}

	action(message, args, mentions) {
		return new Promise(function ($return, $error) {
			let error = false;
			const err = {
				NO_GROUP: 'You must provide a command group to disable.',
				NO_EXIST: `Command group ${ args[0] } does not exist.`,
				DISABLED: `Command group ${ args[0] } is already disabled or is not allowed to be disabled.`
			};

			const self = this; // eslint-disable-line
			function sendError(text) {
				self.respond(message, `**${ text }**`).then(response => response.delete(5 * 1000));
				error = true;
			}

			if (!args[0]) sendError(err.NO_GROUP);else if (!this.bot.commands.groups.includes(args[0])) sendError(err.NO_EXIST);else if (args[0] === 'base' || this.bot.guildStorages.get(message.guild).getSetting('disabledGroups').includes(args[0])) sendError(err.DISABLED);
			if (error) return $return();

			let disabledGroups = this.bot.guildStorages.get(message.guild).getSetting('disabledGroups');
			disabledGroups.push(args[0]);
			this.bot.guildStorages.get(message.guild).setSetting('disabledGroups', disabledGroups);

			this._respond(message, `**Disabled command group "${ args[0] }"**`).then(response => response.delete(5 * 1000));
			return $return();
		}.$asyncbind(this));
	}
}
exports.default = DisableGroup;
