// yamdbf/bin/index.js
'use babel';
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.version = exports.Util = exports.CommandDispatcher = exports.CommandRegistry = exports.CommandLoader = exports.Command = exports.LocalStorage = exports.GuildStorageRegistry = exports.GuildStorageLoader = exports.GuildStorage = exports.Bot = undefined;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _Bot2 = require('./lib/bot/Bot');

var _Bot3 = _interopRequireDefault(_Bot2);

var _GuildStorage2 = require('./lib/storage/GuildStorage');

var _GuildStorage3 = _interopRequireDefault(_GuildStorage2);

var _GuildStorageLoader2 = require('./lib/storage/GuildStorageLoader');

var _GuildStorageLoader3 = _interopRequireDefault(_GuildStorageLoader2);

var _GuildStorageRegistry2 = require('./lib/storage/GuildStorageRegistry');

var _GuildStorageRegistry3 = _interopRequireDefault(_GuildStorageRegistry2);

var _LocalStorage2 = require('./lib/storage/LocalStorage');

var _LocalStorage3 = _interopRequireDefault(_LocalStorage2);

var _Command2 = require('./lib/command/Command');

var _Command3 = _interopRequireDefault(_Command2);

var _CommandLoader2 = require('./lib/command/CommandLoader');

var _CommandLoader3 = _interopRequireDefault(_CommandLoader2);

var _CommandRegistry2 = require('./lib/command/CommandRegistry');

var _CommandRegistry3 = _interopRequireDefault(_CommandRegistry2);

var _CommandDispatcher2 = require('./lib/command/CommandDispatcher');

var _CommandDispatcher3 = _interopRequireDefault(_CommandDispatcher2);

var _Util2 = require('./lib/Util');

var _Util3 = _interopRequireDefault(_Util2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** @exports Bot */
const Bot = exports.Bot = _Bot3.default;
exports.default = Bot;

/** @exports GuildStorage */

const GuildStorage = exports.GuildStorage = _GuildStorage3.default;

/** @exports GuildStorageLoader */
const GuildStorageLoader = exports.GuildStorageLoader = _GuildStorageLoader3.default;

/** @exports GuildStorageRegistry */
const GuildStorageRegistry = exports.GuildStorageRegistry = _GuildStorageRegistry3.default;

/** @exports LocalStorage */
const LocalStorage = exports.LocalStorage = _LocalStorage3.default;

/** @exports Command */
const Command = exports.Command = _Command3.default;

/** @exports CommandLoader */
const CommandLoader = exports.CommandLoader = _CommandLoader3.default;

/** @exports CommandRegistry */
const CommandRegistry = exports.CommandRegistry = _CommandRegistry3.default;

/** @exports CommandDispatcher */
const CommandDispatcher = exports.CommandDispatcher = _CommandDispatcher3.default;

/** @exports Util */
const Util = exports.Util = _Util3.default;

/** @exports version */
const version = exports.version = require(_path2.default.join(__dirname, '..', 'package')).version;

/** @external {Client} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/class/Client} */
/** @external {ClientOptions} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/typedef/ClientOptions} */
/** @external {Collection} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/class/Collection} */
/** @external {Guild} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/class/Guild} */
/** @external {Message} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/class/Message} */
/** @external {User} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/class/User} */
/** @external {PermissionResolvable} See: {@link http://hydrabolt.github.io/discord.js/#!/docs/tag/master/typedef/PermissionResolvable} */
